client/authority-discovery/src/worker.rs

```diff
@@ -610,15 +641,15 @@ where
 			}
 		}
 
-		// Request addresses of authorities.
+		// Request addresses of authorities, refilling the pending lookups queue.
 		if let Poll::Ready(_) = self.query_interval.poll_next_unpin(cx) {
 			// Register waker of underlying task for next interval.
 			while let Poll::Ready(_) = self.query_interval.poll_next_unpin(cx) {}
```

Comment:
This is peculiar, I didn't see this `Interval` implementation via a `Box`ed `Stream` before. Out of curiosity, why not use a single `Delay` per interval with `Delay::reset()` whenever it fires?

Reply:
This dates back to `futures-timer` dropping `Interval` (https://github.com/async-rs/futures-timer/issues/35). We were using `Interval` before and adding our own `Interval` seemed like the best upgrade path at the time.

I do prefer the `Interval` stream interface over the `Delay` interface, as there is no need for the additional `reset` step and thus one way less to misuse it. I am not in favor of authority discovery having its own `Interval` implementation.

Do you see a particular problem with our implementation / usage @romanb?

Reply:
> Do you see a particular problem with our implementation / usage @romanb?

Well, it seems unnecessarily complicated, especially having to do "Register waker of underlying task for next interval." after each tick. Even the original `Interval` in `futures-timer` seems to have just been based on a [`Delay` that is `reset`](https://docs.rs/futures-timer/0.1.1/src/futures_timer/interval.rs.html#61-69) with just some extra calculation to account for skipped intervals that many use-cases probably don't even require.

Reply:
> especially having to do "Register waker of underlying task for next interval." after each tick.

I might be missing something, but wouldn't you have to do the same for a `Delay` after resetting it? Otherwise the underlying task would not be woken up once it is firing, no?

Reply:
Yes, fair point, it just doesn't always need to be so explicit if the periodic work [can be done in a loop](https://github.com/paritytech/substrate/blob/7af8bd048bf9f123465dc9854ab07f05b49f2569/client/service/src/metrics.rs#L180-L209) but that isn't the case here.